pkg/splunk/enterprise/clustermanager.go

```diff
@@ -434,3 +451,160 @@ func VerifyCMisMultisite(ctx context.Context, cr *enterpriseApi.ClusterManager,
 	}
 	return extraEnv, err
 }
+
+// upgradeScenario checks if it is suitable to update the clusterManager based on the Status of the licenseManager, returns bool, err accordingly
+func upgradeScenario(ctx context.Context, c splcommon.ControllerClient, cr *enterpriseApi.ClusterManager) (bool, error) {
+
+	reqLogger := log.FromContext(ctx)
+	scopedLog := reqLogger.WithName("upgradeScenario").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+	eventPublisher, _ := newK8EventPublisher(c, cr)
+
+	namespacedName := types.NamespacedName{
+		Namespace: cr.GetNamespace(),
+		Name:      GetSplunkStatefulsetName(SplunkClusterManager, cr.GetName()),
+	}
+
+	// check if the stateful set is created at this instance
+	statefulSet := &appsv1.StatefulSet{}
+	err := c.Get(ctx, namespacedName, statefulSet)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return true, nil
+	}
+
+	licenseManagerRef := cr.Spec.LicenseManagerRef
+	namespacedName = types.NamespacedName{Namespace: cr.GetNamespace(), Name: licenseManagerRef.Name}
+
+	// create new object
+	licenseManager := &enterpriseApi.LicenseManager{}
+
+	// get the license manager referred in cluster manager
+	err = c.Get(ctx, namespacedName, licenseManager)
+	if err != nil {
+		return true, nil
+	}
+
+	lmImage, err := getLicenseManagerCurrentImage(ctx, c, licenseManager)
+	if err != nil {
+		eventPublisher.Warning(ctx, "upgradeScenario", fmt.Sprintf("Could not get the License Manager Image. Reason %v", err))
+		scopedLog.Error(err, "Unable to licenseManager current image")
+		return false, err
+	}
+	cmImage, err := getClusterManagerCurrentImage(ctx, c, cr)
+	if err != nil {
+		eventPublisher.Warning(ctx, "upgradeScenario", fmt.Sprintf("Could not get the Cluster Manager Image. Reason %v", err))
+		scopedLog.Error(err, "Unable to clusterManager current image")
+		return false, err
+	}
+
+	// check conditions for upgrade
+	if cr.Spec.Image != cmImage && lmImage == cr.Spec.Image && licenseManager.Status.Phase == enterpriseApi.PhaseReady {
+		return true, nil
+	}
+
+	// Temporary workaround to keep the clusterManager method working only when the LM is ready
+	if licenseManager.Status.Phase == enterpriseApi.PhaseReady {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// getClusterManagerCurrentImage gets the image of the pods of the clusterManager before any upgrade takes place,
+// returns the image, and error if something goes wring
+func getClusterManagerCurrentImage(ctx context.Context, c splcommon.ControllerClient, cr *enterpriseApi.ClusterManager) (string, error) {
```

Comment:
Is it possible to write a generic API for all this instead of individual APIs such as getClusterManagerCurrentImage and getLicenseManagerCurrentImage. I see almost everything is common in both the APIs, with the only exception of instanceType(i.e. SplunkClusterManager and SplunkLicenseManager) and `cr`.

Reply:
example here 
```
// getClusterManagerCurrentImage gets the image of the pods of the clusterManager before any upgrade takes place,
// returns the image, and error if something goes wring
func getCRCurrentImage (ctx context.Context, c splcommon.ControllerClient, objectMeta metav1.ObjectMeta ) (string, error) {

	reqLogger := log.FromContext(ctx)
	scopedLog := reqLogger.WithName("getClusterManagerCurrentImage").WithValues("name", objectMeta.GetName(), "namespace", objectMeta.GetNamespace())

	namespacedName := types.NamespacedName{
		Namespace: objectMeta.GetNamespace(),
		Name:      GetSplunkStatefulsetName(SplunkClusterManager, objectMeta.GetName()),
	}
	statefulSet := &appsv1.StatefulSet{}
	err := c.Get(ctx, namespacedName, statefulSet)
	if err != nil {
		scopedLog.Error(err, "StatefulSet types not found in namespace", "namsespace", objectMeta.GetNamespace())
		return "", err
	}
	labelSelector, err := metav1.LabelSelectorAsSelector(statefulSet.Spec.Selector)
	if err != nil {
		scopedLog.Error(err, "Unable to get labels")
		return "", err
	}

	// get a list of all pods in the namespace with matching labels as the statefulset
	statefulsetPods := &corev1.PodList{}
	opts := []rclient.ListOption{
		rclient.InNamespace(objectMeta.GetNamespace()),
		rclient.MatchingLabelsSelector{Selector: labelSelector},
	}

	err = c.List(ctx, statefulsetPods, opts...)
	if err != nil {
		scopedLog.Error(err, "Pods types not found in namespace", "namsespace", objectMeta.GetNamespace())
		return "", err
	}

	// find the container with the phrase 'splunk' in it
	for _, v := range statefulsetPods.Items {
		for _, container := range v.Status.ContainerStatuses {
			if strings.Contains(container.Name, "splunk") {
				image := container.Image
				return image, nil
			}

		}
	}

	return "", nil
}
```

and then call 
```cmImage, err := getCRCurrentImage(ctx, c, &cr.ObjectMeta)```